src/lib/Serialize.hs

```diff
@@ -104,10 +117,81 @@ prettyVal val = case val of
     return $ align $ group innerDoc
   atom -> return $ prettyPrec atom LowestPrec
 
+-- === taking memory snapshots for serializing heap-backed dex values ===
+
+data WithSnapshot a = WithSnapshot a [PtrSnapshot]  deriving Generic
+type RawPtr = Ptr ()
+-- TODO: we could consider using some mmap-able instead of ByteString
+data PtrSnapshot = ByteArray BS.ByteString
+                 | PtrArray [PtrSnapshot]
+                 | NullPtr  deriving Generic
+
+class HasPtrs a where
+  traversePtrs :: Applicative f => (PtrType -> RawPtr -> f RawPtr) -> a -> f a
+
+takeSnapshot :: HasPtrs a => a -> IO (WithSnapshot a)
+takeSnapshot x =
+  liftM (WithSnapshot x) $ execWriterT $ flip traversePtrs x \ptrTy ptrVal -> do
+    snapshot <- lift $ takePtrSnapshot ptrTy ptrVal
+    tell [snapshot]
```

Comment:
This is the classic `Writer []` problem. Maybe use state for now, for the sake of O(n) runtime? There can be a whole bunch of pointers in modules so I'm a little afraid of quadratics here.

Reply:
Sure. Though I'll note that we use `Writer []` a *lot* and this is far from the most egregious use. With laziness, the asymptotic aren't so clear. I wish we had a data structure for writing list-like things with clearly good asymptotics. Happy to switch to state here for now.